### DISCUSSION TITLE
difficulty getting correct tcp checksum..

### DISCUSSION THREAD

**Discussion by livelykitten** [ANLAB-KAIST/KENSv3#169] [Project Q&A 2. Transport Layer (TCP)] [answered]
![image](https://github.com/ANLAB-KAIST/KENSv3/assets/58162598/00c5d7ab-a4f1-4f0f-9c7c-c90219af5bd6)

```
struct tcphdr tcpHdr{}; // zero-initialization
// ...filling in fields except .check
tcpHdr.check = NetworkUtil::tcp_sum(local_ip, peer_ip, (const uint8_t *)&tcpHdr, sizeof(struct tcphdr));
```
This is the code I used for computing tcp checksum, after zero-initializing and filling in struct tcphdr
local_ip and peer_ip are in network byte-order. 

I don't understand why I am getting wrong checksum
could you help me out?

**Comment by Pusnow** [comment, marked as answer]
The `tcp_sum` is used to validate whether a packet has a correct checksum or not.

To fill the checksum field in TCP header, you should use bit-wise not operator.
Also you should follow byte order for the field.

In summary, you should add code like below:

```c++
tcpHdr.check = htons(~tcpHdr.check)
```

BTW, thank you for pointing this out. It will be very difficult for many students.

**Reply by livelykitten** [reply]
Thank you for the quick answer!!! :D I was losing my mind
